### PR TITLE
dask.array: Allow shape to be a numpy array

### LINKED ISSUE #3710: Cannot specify dask array shape with a numpy array
```python
import numpy as np
import dask.array as da


chunks = (10, 10)
shape = np.asarray((100, 100))

n = np.zeros(shape)
d = da.zeros(shape, chunks=chunks)
```

output
```python
Traceback (most recent call last):

  File "<ipython-input-31-0eb572d2c4ad>", line 1, in <module>
    d = da.zeros(shape, chunks=chunks)

  File "/home/mark/miniconda3/envs/owl/lib/python3.6/site-packages/dask/array/wrap.py", line 36, in wrap_func_shape_as_first_arg
    chunks = normalize_chunks(chunks, shape, dtype=dtype)

  File "/home/mark/miniconda3/envs/owl/lib/python3.6/site-packages/dask/array/core.py", line 1990, in normalize_chunks
    if not (len(shape) == 1 and sum(chunks) == shape[0]):

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

You can specify the numpy array shape with a numpy array.

Dask version: 0.18.1
Python: 3.6
Linux 64 bit
Using miniconda

### PR DESCRIPTION
**Pull request by hmaarrfk**
Fixes #3710

~~Still needs tests. Where should put them? I feel like I can parameterize an existing test, but I couldn't find the appropriate one.~~

- [x] Tests added / passed
- [x] Passes `flake8 dask`

### CODE REVIEW

**Comment by hmaarrfk** [comment]
I'm inferring from the name of the function: `wrap_func_shape_as_first_arg`  that it actually takes care of things like shape. Maybe I should apply the same fix to chunks?

I modified the `test_arr_like`  to also test for `empty`, `ones`, `full`, and `zeros`. Maybe that test should be renamed.

I can also make the tests cast chunks to the appropriate type if we think this fix should be applied to chunks as well.

**Comment by hmaarrfk** [comment]
Also allows chunks to be numpy arrays. Though, this is contained in a single commit and can be removed if it is deemed not to be an "bug".

In adding this feature, I noticed that the check for `None` is sometimes replaced by checking if the tuple or list is `Falsy`. Unfortunately, empty numpy arrays are not defined to be Falsy or truthy.

If a desired behaviour is to be able to pass `( ,)` or `[]` to `chunks`, then this would break things as I now only check for `None` and not for `Falsy` `lists` or `tuples`.

File: dask/array/wrap.py [hunk 1]
```diff
@@ -24,6 +24,9 @@ def wrap_func_shape_as_first_arg(func, *args, **kwargs):
     else:
         shape = kwargs.pop('shape')
 
+    if isinstance(shape, np.ndarray):
+        shape = shape.tolist()
```
[the remark this answers is not in the document]
Comment (hmaarrfk) [reply dask/array/wrap.py]: I originally had that, but then there are a few places where tests like 
```
if shape:
```
exist. Empty tuples and shapes are falsy. but numpy arrays are not. The logic in normalized chunks was already really convoluted i felt like that was probably easiest.

**Comment by mrocklin** [comment]
OK, makes sense to me

File: dask/array/tests/test_creation.py [hunk 2]
```diff
@@ -11,27 +11,33 @@
...
     ]
 )
+@pytest.mark.parametrize("cast_shape", [tuple, list, np.asarray])
+@pytest.mark.parametrize("cast_chunks", [tuple, list, np.asarray])
 @pytest.mark.parametrize(
     "shape, chunks", [
-        ((10, 10), (4, 4)),
+        ((10, )    , (4, )    ),
+        ((10, ) * 2, (4, ) * 2),
+        ((10, ) * 6, (4, ) * 6),
```
Comment (hmaarrfk) [inline dask/array/tests/test_creation.py]: Test time increase was due to this line.

Comment (hmaarrfk) [reply dask/array/tests/test_creation.py]: The 6D array creation.

File: dask/array/tests/test_creation.py [hunk 3]
```diff
@@ -11,27 +11,33 @@
...
     ]
 )
 @pytest.mark.parametrize(
     "dtype", [
         "i4",
     ]
 )
-def test_arr_like(funcname, shape, dtype, chunks):
+def test_arr_like(funcname, shape, cast_shape, dtype, cast_chunks, chunks):
```
[the remark this answers is not in the document]
Comment (hmaarrfk) [reply dask/array/tests/test_creation.py]: Good catch.

[the remark this answers is not in the document]
Comment (hmaarrfk) [reply dask/array/tests/test_creation.py]: Github hid my previous comment, it was due to the 6D array I was testing for. Lesson learned.

**Comment by mrocklin** [comment]
OK.  Merging in.  Thanks @hmaarrfk !

**Pull request merged**